### PR TITLE
feat: client src port available in client.connected event

### PR DESCRIPTION
**Pull request by sh3d2**
Fixes https://github.com/emqx/emqx/issues/4231

**If your build fails** due to your commit message not passing the build checks, please review the guidelines here: https://github.com/emqx/emqx/blob/master/CONTRIBUTING.md.

## PR Checklist
Please convert it to a draft if any of the following conditions are not met. Reviewers may skip over until all the items are checked:

- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] In case of non-backward compatible changes, reviewer should check this item as a write-off, and add details in **Backward Compatibility** section

## More information

### CODE REVIEW

**Comment by sh3d2** [comment]
i guess all remaining `client/session` event should have that info. if the way i did it in this PR is ok, i will go with the other events (and webhooks?)

**Review by zmstone — COMMENTED** [review]
Maybe it's better to add `peerport` to `clientinfo()` when `ClientInfo` is constructed.
https://github.com/emqx/emqx/blob/609a8566c7ffb12d8ea6b7381af7b28096904151/src/emqx_channel.erl

Notes:

* Places where `ClientInfo` is constructed (maybe incomplete):
`src/emqx_channel.erl`
`apps/emqx_lwm2m/src/emqx_lwm2m_protocol.erl`
`apps/emqx_exproto/src/emqx_exproto_channel.erl`

* `peername` could be `inet:returned_non_ip_address()`
in which case the second element of the tuple may not be a number, we can use `0` to avoid formatting surprises

**Comment by sh3d2** [comment]
> * `peername` could be `inet:returned_non_ip_address()`
>   in which case the second element of the tuple may not be a number, we can use `0` to avoid formatting surprises

stupid question: 

How do you compare to a non-primitive type (`returned_non_ip_address`)? is there any "quick" way of doing it? (its not possible with guards, is it?)

**Comment by HJianBo** [comment]
> Maybe it's better to add `peerport` to `clientinfo()` when `ClientInfo` is constructed.
> https://github.com/emqx/emqx/blob/609a8566c7ffb12d8ea6b7381af7b28096904151/src/emqx_channel.erl
> 
> Notes:
> 
> * Places where `ClientInfo` is constructed (maybe incomplete):
>   `src/emqx_channel.erl`
>   `apps/emqx_lwm2m/src/emqx_lwm2m_protocol.erl`
>   `apps/emqx_exproto/src/emqx_exproto_channel.erl`
> * `peername` could be `inet:returned_non_ip_address()`
>   in which case the second element of the tuple may not be a number, we can use `0` to avoid formatting surprises


@zmstone  If so, then the following places would be better to add as well?

https://github.com/emqx/emqx/blob/609a8566c7ffb12d8ea6b7381af7b28096904151/apps/emqx_exhook/priv/protos/exhook.proto#L317

https://github.com/emqx/emqx/blob/609a8566c7ffb12d8ea6b7381af7b28096904151/apps/emqx_exhook/src/emqx_exhook_handler.erl#L260

https://github.com/emqx/emqx/blob/609a8566c7ffb12d8ea6b7381af7b28096904151/apps/emqx_exproto/src/emqx_exproto_channel.erl#L579

File: apps/emqx_web_hook/src/emqx_web_hook.erl [hunk 1]
```diff
@@ -117,7 +117,7 @@ on_client_connack(ConnInfo = #{clientid := ClientId, username := Username, peern
 %% Client connected
 %%--------------------------------------------------------------------
 
-on_client_connected(#{clientid := ClientId, username := Username, peerhost := Peerhost}, ConnInfo, _Env) ->
+on_client_connected(#{clientid := ClientId, username := Username, peerhost := Peerhost, peerport:= Peerport}, ConnInfo, _Env) ->
```
Comment (HJianBo) [inline apps/emqx_web_hook/src/emqx_web_hook.erl:120]: @zmstone 
This line needs to be used with care to match `peerport`, which requires all other protocol plugins to pass the `peerport` field in ClientInfo, including some of the Enterprise Edition plugins.

btw,  we also need to update the `clientinfo` declaration  in  the`emqx_types.hrl` file

File: apps/emqx_exproto/src/emqx_exproto_channel.erl [hunk 2]
```diff
@@ -576,17 +576,22 @@ default_conninfo(ConnInfo) ->
...
               expiry_interval => 0}.
 
-default_clientinfo(#{peername := {PeerHost, _},
-                     sockname := {_, SockPort}}) ->
+default_clientinfo(#{peername := {PeerHost, PeerPort},
+    sockname := {_, SockPort}}) when is_integer(PeerPort) ->
     #{zone         => external,
       protocol     => undefined,
       peerhost     => PeerHost,
+      peerport     => PeerPort,
```
Comment (zmstone) [inline apps/emqx_exproto/src/emqx_exproto_channel.erl:584]: maybe add a small help function `peer_port/1` and call it here
so there is no need for another function clause for `default_clientinfo`

```
peer_port(Port) when is_integer(Port) -> Port;
peer_port(_) -> 0.
```
```suggestion
      peerport     => peer_port(PeerPort),
```

Comment (zmstone) [reply apps/emqx_exproto/src/emqx_exproto_channel.erl:584]: and it applies to all other places where `peerport` is initialised.

File: apps/emqx_web_hook/src/emqx_web_hook.erl (hunk 1 above)
Comment (zmstone) [reply apps/emqx_web_hook/src/emqx_web_hook.erl:120]: agree.
we should use `maps:get(peerport, ClientInfo, 0)` to be on the safe side.

the type spec update in emqx_types.erl should give us some level of type check
better than sorry, before we have dialyzer checks everywhere.

**Comment by id** [comment]
It's been too long, closing.

**Pull request closed without merging**